components/resolved_ts/src/observer.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::cell::RefCell;
+use std::sync::Arc;
+
+use engine_traits::KvEngine;
+use kvproto::metapb::{Peer, Region};
+use raft::StateRole;
+use raftstore::coprocessor::*;
+use raftstore::store::fsm::ObserveID;
+use raftstore::store::RegionSnapshot;
+use tikv_util::worker::Scheduler;
+
+use crate::endpoint::Task;
+
+pub struct ChangeDataObserver<E: KvEngine> {
+    cmd_batches: RefCell<Vec<CmdBatch>>,
+    scheduler: Scheduler<Task<E::Snapshot>>,
+}
+
+impl<E: KvEngine> ChangeDataObserver<E> {
+    pub fn new(scheduler: Scheduler<Task<E::Snapshot>>) -> Self {
+        ChangeDataObserver {
+            cmd_batches: RefCell::default(),
+            scheduler,
+        }
+    }
+}
+
+impl<E: KvEngine> Coprocessor for ChangeDataObserver<E> {}
+
+impl<E: KvEngine> CmdObserver<E> for ChangeDataObserver<E> {
+    fn on_prepare_for_apply(&self, observe_id: ObserveID, region_id: u64) {
+        self.cmd_batches
+            .borrow_mut()
+            .push(CmdBatch::new(observe_id, region_id));
+    }
+
+    fn on_apply_cmd(&self, observe_id: ObserveID, region_id: u64, cmd: Cmd) {
+        self.cmd_batches
+            .borrow_mut()
+            .last_mut()
+            .expect("should exist some cmd batch")
+            .push(observe_id, region_id, cmd);
+    }
+
+    fn on_flush_apply(&self, engine: E) {
+        if !self.cmd_batches.borrow().is_empty() {
+            let batches = self.cmd_batches.replace(Vec::default());
+            let mut region = Region::default();
+            region.mut_peers().push(Peer::default());
+            // Create a snapshot here for preventing the old value was GC-ed.
+            // TODO: only need it after enabling old value, may add a flag to indicate whether to get it.
+            let snapshot =
+                RegionSnapshot::from_snapshot(Arc::new(engine.snapshot()), Arc::new(region));
+            if let Err(e) = self.scheduler.schedule(Task::ChangeLog {
```

Comment:
Is there any test covering it?

Reply:
Will cover it in later integration tests